src/masking/mod.rs

```diff
@@ -1 +1,88 @@
+use core::mem::MaybeUninit;
+use core::num::{NonZeroUsize, Wrapping};
+
+use crate::capacity::MaskingCapacity;
+use crate::storage::{PartialStorage, Storage};
+
 mod tests;
+
+pub struct MaskingRingBuffer<S: PartialStorage<Capacity = MaskingCapacity>> {
+    /// The start of the buffer in the storage (`0..CAPACITY`)
+    index: usize,
+    /// The number of items in the buffer (`0..=CAPACITY`)
+    len: usize,
+    /// The underlying storage
+    storage: MaybeUninit<S>,
+}
+
+impl<S: PartialStorage<Capacity = MaskingCapacity>> MaskingRingBuffer<S> {
+    pub fn new(storage: S) -> Self {
+        MaskingRingBuffer {
+            storage: MaybeUninit::new(storage),
+            index: 0,
+            len: 0,
+        }
+    }
+
+    /// Returns whether the ringbuffer is full
+    ///
+    /// A ringbuffer is full when its length equals its capacity. If an item is enqueued
+    /// while the ringbuffer is full [MaskingRingBuffer::enqueue] will dequeue an item to
+    /// make room for the new item. The dequeued item will be returned.
+    pub fn is_full(&self) -> bool {
+        self.len == self.capacity().get()
+    }
+
+    /// Returns true when the ringbuffer is empty
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// The capacity of the underlying storage
+    ///
+    /// This is the maximum number of items that the ringbuffer can hold.
+    pub fn capacity(&self) -> NonZeroUsize {
```

Comment:
Shouldn't we return a `PowerOfTwoCapacity` here, since that's the "true" capacity type?

Reply:
Sorry, `MaskingCapacity`.

Reply:
I made this primarily to compare with the length. I think that the capacity that users will care about is simply the number of elements (maybe even `usize` makes more sense). I like the idea of constructing a ring buffer with certain characteristics and then not having to think about the internals too much anymore.

Reply:
Yeah, as external API, the different capacity types will just get in the way.  I think we should just return `usize` here.

Reply:
One `usize` coming up!